sw/host/opentitanlib/src/bootstrap/rescue.rs

```diff
@@ -0,0 +1,352 @@
+                            if let Ok(_) = self.expect_string(&*uart, "escue") {
+                                eprintln!("c: Entered rescue mode!");
+                                self.synchronize(&*uart)?;
+                                return Ok(());
+                            }
+                        }
+                    }
+                }
+            }
+            eprintln!(" Failed to enter rescue mode.");
+        }
+        bail!(RescueError::RepeatedErrors);
+    }
+}
+
+impl UpdateProtocol for Rescue {
+    fn verify_capabilities(
+        &self,
+        _container: &Bootstrap,
+        transport: &TransportWrapper,
+    ) -> Result<()> {
+        transport
+            .capabilities()
+            .request(Capability::GPIO | Capability::UART)
+            .ok()?;
+        Ok(())
+    }
+
+    /// Returns false, in order to as the containing Bootstrap struct to not perform standard
+    /// BOOTSTRAP/RESET sequence.
+    fn uses_common_bootstrap_reset(&self) -> bool {
+        false
+    }
+
+    /// Performs the update protocol using the `transport` with the firmware `payload`.
+    fn update(
+        &self,
+        container: &Bootstrap,
+        transport: &TransportWrapper,
+        payload: &[u8],
+    ) -> Result<()> {
+        let frames = Frame::from_payload(payload)?;
+        let uart = container.uart_params.create(transport)?;
+
+        self.enter_rescue_mode(container, &*uart)?;
+
+        // Send frames one at a time.
+        'next_block: for transmit_index in 0..frames.len() {
+            let frame = &frames[transmit_index];
+            for consecutive_errors in 0..Self::MAX_CONSECUTIVE_ERRORS {
+                eprint!("{}.", transmit_index);
+                uart.write(frame.as_bytes())?;
+                let mut response = [0u8; Frame::HASH_LEN];
+                let mut index = 0;
+                while index < Frame::HASH_LEN {
+                    let timeout = if index == 0 {
+                        Duration::from_millis(1000)
+                    } else {
+                        Duration::from_millis(10)
+                    };
+                    match uart.read_timeout(&mut response[index..], timeout) {
+                        Ok(0) | Err(_) => break,
+                        Ok(n) => index += n,
+                    }
+                }
+                if index < Frame::HASH_LEN {
+                    eprint!("sync.");
+                    self.synchronize(&*uart)?;
+                    continue;
+                }
+                let response_chunks = response.chunks(4).collect::<Vec<_>>();
+                if response_chunks.iter().all(|&x| x == response_chunks[0]) {
+                    eprint!("sync.");
+                    self.synchronize(&*uart)?;
+                } else if response == frame.frame_hash() {
+                    continue 'next_block;
+                } else {
+                    self.flush_rx(&*uart);
+                    if consecutive_errors >= 3 {
```

Comment:
How is this constant related to MAX_CONSECUTIVE_ERRORS?

Reply:
It is related in that some time before reaching the maximum number of retries, we want to take an additional measure, which it a little time consuming, but could resolve the problem.  I have added a named constant with some comment.